File: health.py
#!/usr/bin/python

import time, random, math, sys
import human_curl as hurl
from optparse import OptionParser

usage = "usage: health.py [options] arg1 arg2"
parser = OptionParser(usage=usage)
parser.add_option("--debug",
                  dest="debug", default=False,
                  help="output all attempted health checks [default: false]")
parser.add_option("--prefix",
                  dest="prefix", default="vulcand",
                  help="prefix of the etcd keyspace used by vulcand [default: vulcand]")
parser.add_option("--site-label",
                  dest="site",
                  help="label used to identify the site's backends and frontends")
parser.add_option("--backend-name",
                  dest="name",
                  help="identifier used for this instance of the backend app")
parser.add_option("--interval",
                  dest="interval", default=30,
                  help="how often to trigger the health check [default: 30 seconds]")
parser.add_option("--etcd",
                  dest="etcdAddress", default="http://127.0.0.1:4001",
                  help="address of the etcd cluster [default: http://120.0.0.1:4001]")
parser.add_option("--backend-address",
                  dest="targetAddress",
                  help="address of the backend to be health checked")
(options, args) = parser.parse_args(args=None, values=None)

if options.debug:
    print '(debug) Parsed arguments:'
    print options

#read args
failCount = 0
failStatus = None
currentBackoff = float(options.interval)
inService = False
address = 'http://httpbin.org/status/200'

#execute health check
def checkHealth():
    global options
    global failCount
    global lastCode
    r = hurl.get(
        str(options.targetAddress),
        allow_redirects=True
    )
    lastCode = r.status_code
    if r.status_code // 100 == 4:
         return False
    elif r.status_code // 100 == 5:
         return False
    else:
         return True

#initialize site
def initializeSite():
    global options
    
    #init backend
    try:
        etcdPath = '%s/v2/keys/%s/backends/%s/backend' % (str(options.etcdAddress), str(options.prefix), str(options.site))
        r1 = hurl.put(etcdPath, data={'value': '{"Type": "http"}'})
        if r1.status_code // 100 == 2:
            print 'Vulcand backend for %s has been initialized.' % str(options.site)
    except:
        print 'Error: Could not communicate with etcd to initialize site %s.' % str(options.site)
        if options.debug:
             print '(debug) Attempted etcd address is %s' % etcdPath
    
    #init frontend route
    try:
        etcdPath = '%s/v2/keys/%s/frontends/%s/frontend' % (str(options.etcdAddress), str(options.prefix), str(options.site))
        r2 = hurl.put(etcdPath, data={'value': '{"Type": "http", "BackendId": "%s", "Route": "Path(`/`)"}' % str(options.site)})
        if r2.status_code // 100 == 2:
            print 'Vulcand frontend for %s has been initialized.' % str(options.site)
    except:
        print 'Error: Could not communicate with etcd to initialize site %s.' % str(options.site)
        if options.debug:
             print '(debug) Attempted etcd address is %s' % etcdPath

#trigger failure
def triggerFailure():
    global options
    global inService
    global failCount

    try:
        etcdPath = '%s/v2/keys/%s/backends/%s/server/%s' % (str(options.etcdAddress), str(options.prefix), str(options.site), str(options.name))
        r = hurl.delete(etcdPath)
        if r.status_code // 100 == 2:
            inService = False
            print 'Target is unhealthy. Removed from rotation.'
    except:
        print 'Error: Could not communicate with etcd to remove target from rotation.'
        if options.debug:
             print '(debug) Attempted etcd address is %s' % etcdPath

#trigger recovery
def triggerRecovery():
    global options
    global inService

    etcdPath = '%s/v2/keys/%s/backends/%s/server/%s' % (str(options.etcdAddress), str(options.prefix), str(options.site), str(options.name))
    try:
        r = hurl.put(etcdPath, data={'value': '{"URL": "%s"}' % address})
        if r.status_code // 100 == 2:
            inService = True
            print 'Target is healthy. Added to rotation.'
    except:
        print 'Error: Could not communicate with etcd to add target back into rotation.'
        if options.debug:
             print '(debug) Attempted etcd address is %s' % etcdPath

#backoff after failure
def backoff():
    global options
    global failCount
    global currentBackoff
    if failCount >= 1:
        currentBackoff = math.ceil(math.exp(failCount))
    else:
        currentBackoff = float(options.interval)

    time.sleep(currentBackoff)

#trigger init
initializeSite()

#event loop
while True:

    healthy = checkHealth()


    if healthy:
        if inService:
             if options.debug:
                 print '(debug) Healthy: %s returned HTTP %s, next check in %s seconds' % (str(options.targetAddress), lastCode, currentBackoff)
        else:
            #add back to rotation and reset values
            failCount = 0
            currentBackoff = options.interval
            print 'Healthy: %s returned HTTP %s, next check in %s seconds' % (str(options.targetAddress), lastCode, currentBackoff)
            triggerRecovery()
    else:
        print 'Failure: %s returned HTTP %s, backing off %s seconds' % (str(options.targetAddress), lastCode, currentBackoff)
        failCount += 1
        if inService:
            #remove from rotation
            triggerFailure()

    backoff()

File: health.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"time"
)

const (
	checkInterval = time.Second * 30
	maxSleep      = time.Minute * 5
)

type settings struct {
	debug         bool
	prefix        string
	siteName      string
  siteHostname  string
	backendName   string
	interval      int
	targetAddress string
	etcdAddress   string
}

type vulcanBackendInstancePayload struct {
	URL string `json:"URL"`
}

type vulcanBackendPayload struct {
	Type string `json:"Type"`
}

type vulcanFrontendPayload struct {
	Type        string `json:"Type"`
	BackendName string `json:"BackendId"`
	Route       string `json:"Route"`
}

var inService bool
var siteName string
var lastStatusCode int

func main() {
	fs := flag.NewFlagSet("sidekick-flagset", flag.ExitOnError)
	debug := fs.Bool("debug", false, "output debug info and log all attempted health checks")
	prefix := fs.String("prefix", "vulcand", "prefix of the etcd keyspace used by vulcand")
	siteName := fs.String("site-name", "", "label used to identify the site's backends and frontends")
  siteHostname := fs.String("site-hostname", "", "hostname of the site this backend serves")
	backendName := fs.String("backend-name", "", "identifier used for this instance of the backend app")
	interval := fs.Int("interval", 30, "how often to trigger the health check in seconds")
	targetAddress := fs.String("target-address", "", "address of the backend to be health checked")
	etcdAddress := fs.String("etcd-address", "http://localhost:4001", "address of the etcd cluster")

	if err := fs.Parse(os.Args[1:]); err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}

	if *targetAddress == "" {
		fmt.Fprintln(os.Stderr, "target-address is required")
		os.Exit(1)
	}
	//TODO: add these

	s := settings{
		debug:         *debug,
		prefix:        *prefix,
		siteName:      *siteName,
    siteHostname:  *siteHostname,
		backendName:   *backendName,
		interval:      *interval,
		targetAddress: *targetAddress,
		etcdAddress:   *etcdAddress,
	}

	//initialize variables
	healthy := false
	sleep := time.Second * time.Duration(s.interval)

	if s.debug {
		fmt.Printf("(debug) Using settings: %v\n", s)
	}

	//InitializeSite(&s)

	for {
		healthy = HealthCheck(&s)

		if healthy {
			if inService {
				if *debug {
					log.Printf("(debug) Healthy: %v returned HTTP %v, next check in %v seconds", s.targetAddress, lastStatusCode, sleep)
				}
			} else {
				//reset sleep counter
				sleep = time.Second * time.Duration(s.interval)
				log.Printf("Healthy: %v returned HTTP %v, next check in %v seconds", s.targetAddress, lastStatusCode, sleep)
				TriggerRecovery(&s)
			}
		} else {
			//calc backoff
			sleep = ExpBackoff(sleep, maxSleep)
			log.Printf("Failure: %v returned HTTP %v, backing off %v seconds", s.targetAddress, lastStatusCode, sleep)

			if inService {
				TriggerFailure(&s)
			}
		}
		time.Sleep(sleep)
	}
}

func TriggerRecovery(s *settings) {
	//create new etcd client
	ec := &etcdClient{
		Address: s.etcdAddress,
		Prefix:  "v2/keys/" + s.prefix,
	}

	//create our backend instance
	//TODO: abstract out
	p2 := vulcanBackendInstancePayload{
		URL: s.targetAddress,
	}
	backendInstancePath := "/backends/" + s.siteName + "/servers/" + s.backendName
	err2 := ec.Put(backendInstancePath, p2, s)
	if err2 != nil {
		fmt.Printf("error putting value to etcd: %v", err2)
		return
	}
	if s.debug {
		log.Printf("(debug) Vulcand backend instance for %s has been initialized.", s.backendName)
	}

	inService = true
	log.Printf("Target is healthy. Added to rotation.")
}

func TriggerFailure(s *settings) {
	//create new etcd client
	ec := &etcdClient{
		Address: s.etcdAddress,
		Prefix:  "v2/keys/" + s.prefix,
	}

	//delete the backend instance
	backendInstancePath := "/backends/" + s.siteName + "/servers/" + s.backendName
	err := ec.Delete(backendInstancePath, s)
	if err != nil {
		fmt.Printf("error deleting value from etcd: %v", err)
		return
	}

	inService = false
	log.Printf("Target is unhealthy. Removed from rotation.")
}

func InitializeSite(s *settings) {
	//create new etcd client
	ec := &etcdClient{
		Address: s.etcdAddress,
		Prefix:  "v2/keys/" + s.prefix,
	}

	//create our backend
	p1 := vulcanBackendPayload{
		Type: "http",
	}
	backendPath := "/backends/" + s.siteName + "/backend"
	err := ec.Put(backendPath, p1, s)
	if err != nil {
		fmt.Printf("Error creating backend %s: %v", s.siteName, err)
		return
	}
	if s.debug {
		log.Printf("(debug) Vulcand backend for %s has been initialized.", s.siteName)
	}

	//create our frontend instance
  routingRegix := fmt.Sprintf("Host('%s') && PathRegexp('/.*')", s.siteHostname)
	p2 := vulcanFrontendPayload{
		Type:        "http",
		BackendName: s.siteName,
		Route:       routingRegix,
	}
	frontendPath := "/frontends/" + s.siteName + "/frontend"
	err2 := ec.Put(frontendPath, p2, s)
	if err2 != nil {
		fmt.Printf("error putting value to etcd: %v", err2)
		return
	}
	if s.debug {
		log.Printf("(debug) Vulcand frontend for %s (%s) has been initialized.", s.siteName, s.siteHostname)
	}

	log.Printf("Frontend and backend for %s have been initialized.", s.siteName)

}

type etcdClient struct {
	Address string
	Prefix  string
}

func ConstructPut(path string, json string, s *settings) (*http.Response, error) {
  body := bytes.NewBuffer([]byte("value=" + json))

  req, err := http.NewRequest("PUT", path, body)
  if err != nil {
    return nil, err
  }

  req.Header.Add("Content-Type", "application/x-www-form-urlencoded")

  resp, err := http.DefaultClient.Do(req)
  if err != nil {
    return nil, err
  }

  return resp, nil
}

func (c *etcdClient) Put(path string, value interface{}, s *settings) error {

  //stringify json
  json, err := json.Marshal(value)
  if err != nil {
    return err
  }

  //construct the request url
	u := fmt.Sprintf("%s/%s%s", c.Address, c.Prefix, path)

	resp, err := ConstructPut(u, string(json), s)

	if err != nil {
    return err
  }

  if s.debug {
    fmt.Printf("Status code from etcd is HTTP %d", resp.StatusCode)
  }

  if int(resp.StatusCode/100) == 3 {
    location := resp.Header["Location"][0]
    resp, err = ConstructPut(location, string(json), s)

    if err != nil {
      return err
    }
  }

  //check for any 50x
  if int(resp.StatusCode/100) == 5 {
    return fmt.Errorf("etcd unexpectedly returned HTTP %d \n", resp.StatusCode)
  }

  return nil
}

func (c *etcdClient) Delete(path string, s *settings) error {

	u := fmt.Sprintf("%s/%s%s", c.Address, c.Prefix, path)

	req, err := http.NewRequest("DELETE", u, nil)
	if err != nil {
		return err
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}

	defer resp.Body.Close()
	respBody, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	if s.debug {
		fmt.Println(string(respBody))
	}

	//check for any non-20x
	if int(resp.StatusCode/100) == 5 {
		return fmt.Errorf("etcd unexpectedly returned HTTP %d \n", resp.StatusCode)
	}

	return nil
}

func HealthCheck(s *settings) bool {
	resp, err := http.Get(s.targetAddress)
	if err != nil {
		log.Printf("Error checking target %v: %v", s.targetAddress, err)
		return false
	}

	lastStatusCode = resp.StatusCode

	//check for any 20x
	if int(resp.StatusCode/100) == 2 {
		return true
	}

	return false
}

func ExpBackoff(prev, max time.Duration) time.Duration {
	if prev == 0 {
		return time.Second
	}
	if prev > max/2 {
		return max
	}
	return 2 * prev
}
